# Analyze Borrower's Default Risk

Introduction
The idea is to prepare a report for the credit division of a bank. We will find out the effect of a customer's marital status and the number of children he has on the probability of default in repayment of the loan. The bank already has some data regarding the creditworthiness of customers.

The report will be considered when making **credit assessment** for prospective customers. **Credit scoring** is used to evaluate the ability of potential borrowers to repay their loans.


Stages:
1. Identify and fill in missing values
2. Replace the real number data type with integer type
3. Remove duplicate data
4. Categorize data

Purpose:
find out the effect of a customer's marital status and the number of children he has on the probability of default in repaying the loan

The hypothesis tested:
Is there a relationship between having children and the probability of a person defaulting on a loan?
Is there a relationship between marital status and the probability of a person defaulting on a loan?
Is there a relationship between the level of income and the probability of a person defaulting on a loan?
How do different loan objectives affect a person's probability of defaulting on a loan?

In [1]:

import pandas as pd



In [2]:

df = pd.read_csv('/datasets/credit_scoring_eng.csv')

In [3]:

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:

df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [5]:

df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000



total data per kolom 21525, namun diduga terdapat nilai hilang pada kolom 'days_employed' dan 'total_income' yang jumlah datanya 19351

In [6]:


df[df['days_employed'].isna()].head()

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding






after looking at the missing data table and the number of rows it can be surmised if the missing data in the 'days_employed' and 'total_income' columns are symmetrical.

In [7]:


df[(df['days_employed'].isnull()) & (df['total_income'].isnull())].head()

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [8]:


df[(df['days_employed'].isnull()) & (df['total_income'].isnull())].describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

after digging into missing values of the days_employed and total_income columns, we can conclude that both of the missing values are indentical



education column

In [9]:


sorted(df['education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'GRADUATE DEGREE',
 'Graduate Degree',
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

make it into lowercase letter

In [10]:


df['education'] = df['education'].str.lower()

recheck again

In [11]:


sorted(df['education'].unique())

["bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

children column

In [12]:

df['children'].describe()

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64



there is -1 child, but there is also 20 children, which will be examined further


In [13]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [14]:
df[df['children']==20]['dob_years'].unique()

array([21, 44, 56, 59,  0, 35, 23, 40, 26, 45, 37, 24, 64, 69, 46, 54, 55,
       60, 36, 41, 32, 34, 42, 49, 30, 43, 57, 61, 50, 31, 48, 25, 38, 62,
       27, 33, 39, 29, 51, 53, 52])

In [15]:
df[df['children']==20]['family_status'].unique()

array(['married', 'widow / widower', 'unmarried', 'civil partnership',
       'divorced'], dtype=object)

In [16]:
df[df['children']==-1]['dob_years'].unique()

array([46, 50, 57, 54, 55, 38, 42, 26, 41, 64, 32, 48, 34, 51, 35, 53, 31,
       28, 37, 63, 59, 30, 27, 61, 23, 44, 40, 33, 43, 69])

In [17]:
df[df['children']==-1]['family_status'].unique()

array(['civil partnership', 'married', 'unmarried', 'divorced',
       'widow / widower'], dtype=object)

In [18]:
((76+47)/21525)*100

0.5714285714285714

data on the number of 20 children is random, there is even a customer age of 21 years who already has 20 children and when viewed from the family status it also looks random
data on children totaling -1 occurs randomly, it is assumed that the value -1 occurs in unmarried/divorced family status, but is also found in married data
conclusion: data on the number of children 20 and -1 will be deleted because the percentage is also small at 0.5%

In [19]:

df['children'] = df[(df['children'] != -1) & (df['children'] != 20)]

In [20]:

df['children'].value_counts()

0.0    14149
1.0     4818
2.0     2055
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

days_employed column



In [21]:
df['days_employed']

0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64

In [22]:
df[df['days_employed']<0]['income_type'].unique()

array(['employee', 'business', 'civil servant', 'student', 'entrepreneur',
       'paternity / maternity leave'], dtype=object)

In [23]:
df[df['days_employed']>0]['income_type'].unique()

array(['retiree', 'unemployed'], dtype=object)

the days_employed column contains negative data, if you refer to the income_type column, only retirees and unemployed are positive

In [24]:
min(df[(df['days_employed']/365)>51]['days_employed']/365)

900.6266317932007

In [25]:


(len(df[df['days_employed'] > 0])/df['days_employed'].count())*100

17.80269753501111


column days_employed negative working days seen from income_type is a type of productive worker so it is assumed that negative data can be turned positive
Column days_employed positive working days seen from income_type is a type of unproductive worker so it is assumed that positive data can be changed to 0
Column days_employed positive working days seen from dob_years and divided by 365 then the result is so big that it doesn't logically become working days

In [26]:
df['days_employed'] = df['days_employed'].abs()

the negative days in column days_employed will be turned in positive

dob_years column

In [27]:
df.describe()

days_employed     dob_years  education_id  family_status_id  \
count   19351.000000  21525.000000  21525.000000      21525.000000   
mean    66914.728907     43.293380      0.817236          0.972544   
std    139030.880527     12.574584      0.548138          1.420324   
min        24.141633      0.000000      0.000000          0.000000   
25%       927.009265     33.000000      1.000000          0.000000   
50%      2194.220567     42.000000      1.000000          0.000000   
75%      5537.882441     53.000000      1.000000          1.000000   
max    401755.400475     75.000000      4.000000          4.000000   

               debt   total_income  
count  21525.000000   19351.000000  
mean       0.080883   26787.568355  
std        0.272661   16475.450632  
min        0.000000    3306.762000  
25%        0.000000   16488.504500  
50%        0.000000   23202.870000  
75%        0.000000   32549.611000  
max        1.000000  362496.645000

In [28]:
df.head()

children  days_employed  dob_years            education  education_id  \
0      1.0    8437.673028         42    bachelor's degree             0   
1      1.0    4024.803754         36  secondary education             1   
2      0.0    5623.422610         33  secondary education             1   
3      3.0    4124.747207         32  secondary education             1   
4      0.0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [29]:
df[df['dob_years']==0]['family_status'].unique()

array(['married', 'divorced', 'unmarried', 'widow / widower',
       'civil partnership'], dtype=object)

In [30]:
df[df['dob_years']==0]['gender'].unique()

array(['F', 'M'], dtype=object)

In [31]:
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

column dob_years contains age = 0
after reviewing the data looks random and has no correlation with other data and the amount is small then the data will be deleted

In [32]:


(len(df[df['dob_years'] == 0]) / df['dob_years'].count()) * 100

0.4692218350754936



all data whose age = 0 is deleted

In [33]:
df.head()

children  days_employed  dob_years            education  education_id  \
0      1.0    8437.673028         42    bachelor's degree             0   
1      1.0    4024.803754         36  secondary education             1   
2      0.0    5623.422610         33  secondary education             1   
3      3.0    4124.747207         32  secondary education             1   
4      0.0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [34]:


df = df[df['dob_years'] != 0]

In [35]:

df[df['dob_years'] == 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

family_status column

In [36]:


df['family_status'].value_counts()

married              12331
civil partnership     4156
unmarried             2797
divorced              1185
widow / widower        955
Name: family_status, dtype: int64

In [37]:
df['family_status'].describe()

count       21424
unique          5
top       married
freq        12331
Name: family_status, dtype: object

there is 'XNA' gender, but only 1 data and it seems that the relationship is normal with the other data
because 'gender' has nothing to do with the ability to borrow and make money, the data is left as is

In [38]:
df['gender'].value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

In [39]:


df[df['gender'] == 'XNA']

children  days_employed  dob_years     education  education_id  \
10701      0.0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate



income_type column

In [40]:


df['income_type'].value_counts()

employee                       11064
business                        5065
retiree                         3836
civil servant                   1453
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [41]:
df.groupby('income_type')['total_income'].sum()

income_type
business                       1.476996e+08
civil servant                  3.573388e+07
employee                       2.573171e+08
entrepreneur                   7.986610e+04
paternity / maternity leave    8.612661e+03
retiree                        7.516408e+07
student                        1.571226e+04
unemployed                     4.202872e+04
Name: total_income, dtype: float64

In [42]:
df[df['income_type'] == 'unemployed']

children  days_employed  dob_years            education  education_id  \
3133       1.0  337524.466835         31  secondary education             1   
14798      0.0  395302.838654         45    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
3133             married                 0      M  unemployed     1   
14798  civil partnership                 1      F  unemployed     0   

       total_income                          purpose  
3133       9593.119  buying property for renting out  
14798     32435.602               housing renovation

In [43]:
df[df['income_type'] == 'retiree']

children  days_employed  dob_years            education  education_id  \
4          0.0  340266.072047         53  secondary education             1   
12         0.0            NaN         65  secondary education             1   
18         0.0  400281.136913         53  secondary education             1   
24         1.0  338551.952911         57  secondary education             1   
25         0.0  363548.489348         67  secondary education             1   
...        ...            ...        ...                  ...           ...   
21505      0.0  338904.866406         53  secondary education             1   
21508      0.0  386497.714078         62  secondary education             1   
21509      0.0  362161.054124         59    bachelor's degree             0   
21518      0.0  373995.710838         59  secondary education             1   
21521      0.0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      F     retiree     0   
12     civil partnership                 1      M     retiree     0   
18       widow / widower                 2      F     retiree     0   
24             unmarried                 4      F     retiree     0   
25               married                 0      M     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      M     retiree     0   
21508            married                 0      M     retiree     0   
21509            married                 0      M     retiree     0   
21518            married                 0      F     retiree     0   
21521            married                 0      F     retiree     0   

       total_income                                   purpose  
4         25378.572                         to have a wedding  
12              NaN                         to have a wedding  
18         9091.804                  buying a second-hand car  
24        46487.558  transactions with commercial real estate  
25         8818.041                           buy real estate  
...             ...                                       ...  
21505     12070.399                         to have a wedding  
21508     11622.175                                  property  
21509     11684.650                  real estate transactions  
21518     24618.344                         purchase of a car  
21521     24959.969                         purchase of a car  

[3836 rows x 12 columns]

there are some category in income_type column that unemployed and retiree wanted to borrow money, but logically as long as they have money or investment that generate cash, it won't be a problem even if unemployed o retiree, so we will let the data as it is

handling duplicate data

In [44]:


df.duplicated().sum()

71

In [45]:
df['education'].value_counts()

secondary education    15169
bachelor's degree       5225
some college             742
primary education        282
graduate degree            6
Name: education, dtype: int64

In [46]:
df['family_status'].value_counts()

married              12331
civil partnership     4156
unmarried             2797
divorced              1185
widow / widower        955
Name: family_status, dtype: int64

In [47]:
df['income_type'].value_counts()

employee                       11064
business                        5065
retiree                         3836
civil servant                   1453
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

erasing duplicate data

In [48]:


df = df.drop_duplicates()

In [49]:

df.duplicated().sum()

0

checking the data after cleaning process

In [50]:


df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21353 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21231 non-null  object 
 1   days_employed     19260 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      19260 non-null  float64
 11  purpose           21353 non-null  object 
dtypes: float64(2), int64(4), object(6)
memory usage: 2.1+ MB




as a whole except for the days_employed column and total income, the data has decreased from 21525 to 20763 data or reduced by 3.5% data

Handing Missing Values

Missing Values in total_income

In [51]:
sorted(df['dob_years'].unique())

[19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75]

creating function to categorize the age

In [52]:


def age_group(age):
    if age <= 56:
        return 'produktif'
    else:
        return 'non produktif'

In [53]:


age_group(53)

'produktif'

In [54]:


df['age_group'] = df['dob_years'].apply(age_group)

In [55]:

df['age_group']

0            produktif
1            produktif
2            produktif
3            produktif
4            produktif
             ...      
21520        produktif
21521    non produktif
21522        produktif
21523        produktif
21524        produktif
Name: age_group, Length: 21353, dtype: object

checking the details

In [56]:


df1 = df[df['days_employed'].notnull()]

In [57]:
df1.groupby('income_type')['total_income'].sum()

income_type
business                       1.476996e+08
civil servant                  3.573388e+07
employee                       2.573171e+08
entrepreneur                   7.986610e+04
paternity / maternity leave    8.612661e+03
retiree                        7.516408e+07
student                        1.571226e+04
unemployed                     4.202872e+04
Name: total_income, dtype: float64

In [58]:
df1['total_income'].describe()

count     19260.000000
mean      26794.435857
std       16493.972601
min        3306.762000
25%       16494.872500
50%       23201.873500
75%       32539.207750
max      362496.645000
Name: total_income, dtype: float64

In [59]:


df1.groupby('income_type')['total_income'].mean()

income_type
business                       32397.357125
civil servant                  27361.316126
employee                       25824.679592
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21939.310393
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [60]:
df1['total_income'].median()

23201.8735

the type of work that has the most influence on income_type, because there is a prominent income value in several types of work, the median will be used, because the mean of total_income is spread higher so it is less valid

In [61]:
df1['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'student', 'entrepreneur', 'paternity / maternity leave'],
      dtype=object)

creating function to fill missing values

In [62]:


def fill_missing_value(dataframe, agg_column, value_column):
    grouped_values = dataframe.groupby(agg_column)[value_column].median().reset_index()
    size = len(grouped_values)
    for i in range(size):
        group = grouped_values[agg_column][i]
        value = grouped_values[value_column][i]
        dataframe.loc[(dataframe[agg_column]==group) & (dataframe[value_column].isna()), value_column] = value
    return dataframe


In [63]:


fill_missing_value(df, 'income_type', 'total_income')   

children  days_employed  dob_years            education  education_id  \
0          1.0    8437.673028         42    bachelor's degree             0   
1          1.0    4024.803754         36  secondary education             1   
2          0.0    5623.422610         33  secondary education             1   
3          3.0    4124.747207         32  secondary education             1   
4          0.0  340266.072047         53  secondary education             1   
...        ...            ...        ...                  ...           ...   
21520      1.0    4529.316663         43  secondary education             1   
21521      0.0  343937.404131         67  secondary education             1   
21522      1.0    2113.346888         38  secondary education             1   
21523      3.0    3112.481705         38  secondary education             1   
21524      2.0    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose      age_group  
0         40620.102    purchase of the house      produktif  
1         17932.802             car purchase      produktif  
2         23341.752    purchase of the house      produktif  
3         42820.568  supplementary education      produktif  
4         25378.572        to have a wedding      produktif  
...             ...                      ...            ...  
21520     35966.698     housing transactions      produktif  
21521     24959.969        purchase of a car  non produktif  
21522     14347.610                 property      produktif  
21523     39054.888        buying my own car      produktif  
21524     13127.587             to buy a car      produktif  

[21353 rows x 13 columns]

In [64]:


df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21353 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21231 non-null  object 
 1   days_employed     19260 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      21353 non-null  float64
 11  purpose           21353 non-null  object 
 12  age_group         21353 non-null  object 
dtypes: float64(2), int64(4), object(7)
memory usage: 2.3+ MB


Missing Values in days_employed

In [65]:
df1.groupby('income_type')['days_employed'].sum()

income_type
business                       9.632248e+06
civil servant                  4.425392e+06
employee                       2.320221e+07
entrepreneur                   5.208481e+02
paternity / maternity leave    3.296760e+03
retiree                        1.250544e+09
student                        5.787516e+02
unemployed                     7.328273e+05
Name: days_employed, dtype: float64

In [66]:

def fill_missing_value(dataframe, agg_column, value_column):
    grouped_values = dataframe.groupby(agg_column)[value_column].median().reset_index()
    size = len(grouped_values)
    for i in range(size):
        group = grouped_values[agg_column][i]
        value = grouped_values[value_column][i]
        dataframe.loc[(dataframe[agg_column]==group) & (dataframe[value_column].isna()), value_column] = value
    return dataframe

In [67]:


fill_missing_value(df, 'income_type', 'days_employed') 

children  days_employed  dob_years            education  education_id  \
0          1.0    8437.673028         42    bachelor's degree             0   
1          1.0    4024.803754         36  secondary education             1   
2          0.0    5623.422610         33  secondary education             1   
3          3.0    4124.747207         32  secondary education             1   
4          0.0  340266.072047         53  secondary education             1   
...        ...            ...        ...                  ...           ...   
21520      1.0    4529.316663         43  secondary education             1   
21521      0.0  343937.404131         67  secondary education             1   
21522      1.0    2113.346888         38  secondary education             1   
21523      3.0    3112.481705         38  secondary education             1   
21524      2.0    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose      age_group  
0         40620.102    purchase of the house      produktif  
1         17932.802             car purchase      produktif  
2         23341.752    purchase of the house      produktif  
3         42820.568  supplementary education      produktif  
4         25378.572        to have a wedding      produktif  
...             ...                      ...            ...  
21520     35966.698     housing transactions      produktif  
21521     24959.969        purchase of a car  non produktif  
21522     14347.610                 property      produktif  
21523     39054.888        buying my own car      produktif  
21524     13127.587             to buy a car      produktif  

[21353 rows x 13 columns]

In [68]:


df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21353 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21231 non-null  object 
 1   days_employed     21353 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      21353 non-null  float64
 11  purpose           21353 non-null  object 
 12  age_group         21353 non-null  object 
dtypes: float64(2), int64(4), object(7)
memory usage: 2.3+ MB


Categorizing Data


purpose column

In [69]:
df.head()

children  days_employed  dob_years            education  education_id  \
0      1.0    8437.673028         42    bachelor's degree             0   
1      1.0    4024.803754         36  secondary education             1   
2      0.0    5623.422610         33  secondary education             1   
3      3.0    4124.747207         32  secondary education             1   
4      0.0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  age_group  
0    purchase of the house  produktif  
1             car purchase  produktif  
2    purchase of the house  produktif  
3  supplementary education  produktif  
4        to have a wedding  produktif

In [70]:


df['purpose'].value_counts()

wedding ceremony                            786
having a wedding                            764
to have a wedding                           760
real estate transactions                    672
buy commercial real estate                  658
buying property for renting out             649
transactions with commercial real estate    648
housing transactions                        646
purchase of the house                       640
housing                                     640
purchase of the house for my family         637
construction of own property                633
property                                    629
transactions with my real estate            627
building a real estate                      621
purchase of my own house                    619
building a property                         619
buy real estate                             618
housing renovation                          605
buy residential real estate                 603
buying my own car                       

In [71]:


df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

the purpose category is divided into 4 major sections, namely buying cars, education, real estate, and education

In [72]:

def categorize_purp(data):
    wedding = ['wedding']
    education = ['university', 'education', 'educated']
    real_estate = ['real estate', 'property', 'housing', 'commercial', 'house']
    car_need = ['car','cars']
    for wed in wedding:
        if wed in data:
            return 'wedding'
    for car in car_need:
        if car in data:
            return 'car'
    for edu in education:
        if edu in data:
            return 'education'
    for rea in real_estate:
        if rea in data:
            return 'real estate'

In [73]:


df['categorize_purpose'] = df['purpose'].apply(categorize_purp)

In [74]:


df[['days_employed', 'total_income']]

days_employed  total_income
0        8437.673028     40620.102
1        4024.803754     17932.802
2        5623.422610     23341.752
3        4124.747207     42820.568
4      340266.072047     25378.572
...              ...           ...
21520    4529.316663     35966.698
21521  343937.404131     24959.969
21522    2113.346888     14347.610
21523    3112.481705     39054.888
21524    1984.507589     13127.587

[21353 rows x 2 columns]

total_income column

In [75]:
# Dapatkan kesimpulan statistik untuk kolomnya

df['total_income'].describe()

count     21353.000000
mean      26457.685573
std       15727.498629
min        3306.762000
25%       17224.099000
50%       22815.103500
75%       31325.638000
max      362496.645000
Name: total_income, dtype: float64



grouping total income into 3 categories, namely <20000 , 20000<=<30000, >30000

In [76]:


def income_category(x):
    if x <20000:
        return 'low'
    if x >30000:
        return 'high'
    else:
        return 'medium'

In [77]:

df['income_category'] = df['total_income'].apply(income_category)

In [78]:

df.describe()

days_employed     dob_years  education_id  family_status_id  \
count   21353.000000  21353.000000  21353.000000      21353.000000   
mean    67027.721505     43.475905      0.817777          0.972650   
std    139173.908853     12.242316      0.548764          1.420987   
min        24.141633     19.000000      0.000000          0.000000   
25%      1023.701831     33.000000      1.000000          0.000000   
50%      1998.340741     43.000000      1.000000          0.000000   
75%      5320.815418     53.000000      1.000000          1.000000   
max    401755.400475     75.000000      4.000000          4.000000   

               debt   total_income  
count  21353.000000   21353.000000  
mean       0.081160   26457.685573  
std        0.273086   15727.498629  
min        0.000000    3306.762000  
25%        0.000000   17224.099000  
50%        0.000000   22815.103500  
75%        0.000000   31325.638000  
max        1.000000  362496.645000

Hypothesis testing


**Is there a correlation between having children and the probability of defaulting on a loan?**

In [79]:
df[['children', 'debt']]

children  debt
0          1.0     0
1          1.0     0
2          0.0     0
3          3.0     0
4          0.0     0
...        ...   ...
21520      1.0     0
21521      0.0     0
21522      1.0     1
21523      3.0     1
21524      2.0     0

[21353 rows x 2 columns]

In [80]:
df_child = df.groupby('children').agg({'debt':['sum', 'count']})
df_child.columns = df_child.columns.droplevel()
df_child['sum']/df_child['count']

children
0.0    0.075453
1.0    0.092028
2.0    0.095145
3.0    0.082317
4.0    0.097561
5.0    0.000000
dtype: float64

In [81]:


((df_child['sum']/df_child['count'])*100).sort_values().reset_index()

children         0
0       5.0  0.000000
1       0.0  7.545286
2       3.0  8.231707
3       1.0  9.202838
4       2.0  9.514468
5       4.0  9.756098


In conclusion, from the table it appears that the number of children does not affect the percentage of default, where the number of children 4 and the number of children 1 have almost the same percentage of default, namely 9%


**Is there a correlation between family status and the probability of defaulting on a loan?**

In [82]:
df_family = df.groupby('family_status').agg({'debt':['sum', 'count']})
df_family.columns = df_family.columns.droplevel()
df_family['sum']/df_family['count']

family_status
civil partnership    0.093462
divorced             0.071730
married              0.075427
unmarried            0.097709
widow / widower      0.064990
dtype: float64

In [83]:

((df_family['sum']/df_family['count'])*100).sort_values().reset_index()

family_status         0
0    widow / widower  6.498952
1           divorced  7.172996
2            married  7.542718
3  civil partnership  9.346247
4          unmarried  9.770938




the conclusion that is obtained is that there is no correlation between family status and the percentage of default, because at first glance it appears that umarried status has the same percentage of default with civil partnership

**Is there a correlation between income levels and the probability of defaulting on a loan?**

In [84]:
df_income = df.groupby('income_category').agg({'debt':['sum', 'count']})
df_income.columns = df_income.columns.droplevel()
df_income['sum']/df_income['count']

income_category
high      0.073841
low       0.082955
medium    0.084936
dtype: float64

In [85]:


((df_income['sum']/df_income['count'])*100).sort_values().reset_index()

income_category         0
0            high  7.384145
1             low  8.295528
2          medium  8.493610





In conclusion, from the table it appears that there is a correlation between the level of income and the percentage of default, where high income results in a lower percentage of default than medium and low income.

**How do credit goals affect the default rate?**

In [86]:
df_purpose = df.groupby('categorize_purpose').agg({'debt':['sum', 'count']})
df_purpose.columns = df_purpose.columns.droplevel()
df_purpose['sum']/df_purpose['count']

categorize_purpose
car            0.093371
education      0.092616
real estate    0.072371
wedding        0.079654
dtype: float64

In [87]:


((df_purpose['sum']/df_purpose['count'])*100).sort_values().reset_index()

categorize_purpose         0
0        real estate  7.237087
1            wedding  7.965368
2          education  9.261577
3                car  9.337068

in conclusion the purpose of borrowing for car is the highest default rate, followed by education, wedding and real estate

**Final Conclusion**

1. the data is checked in general so that an initial picture is visible, each column is checked whether each data is in accordance with the basic logic and its relation to other data
2. Missing data is filled in by paying attention to the columns that affect it and a decision is made to fill in/delete based on logic
3. Duplicate data is handled with respect to non-numeric columns
4. loans to the real estate category are the loan objectives with the lowest percentage of default, then wedding, education, car, the percentages are increasing successively
5. Loans for the car category, namely for buying a car, are the loan objectives with the highest default percentage